### PR TITLE
Set default page size for some EC2 operations

### PR DESCRIPTION
**Pull request by JordonPhillips**
When customers use max-results on EC2 operations, the entire
response will be returned and then the CLI will filter out what
the customer wants. This can cause significant pauses for what
seems like a simple request. This change will set the default
page size on describe-volumes and describe-snapshots if no
incompatible arguments are specified.

cc @kyleknap @jamesls

### CODE REVIEW

File: awscli/customizations/ec2/paginate.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+
+class EC2PageSizeInjector(object):
+
+    DEFAULT_TARGET_OPERATIONS = {
+        "describe-volumes": [],
+        "describe-snapshots": ['owner-ids', 'restorable_by_user_ids']
```
Comment (kyleknap) [inline awscli/customizations/ec2/paginate.py]: Why does ``owner-ids`` have a hyphen and ``restorable_by_user_ids`` uses underscores?

File: awscli/customizations/ec2/paginate.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    DEFAULT_GLOBAL_WHITELIST = [
+        'cli_input_json', 'generate_cli_skeleton', 'help', 'max_items',
+        'next_token', 'dry_run', 'starting_token'
+    ]
+
+    def __init__(self, default_page_size=1000, target_operations=None,
```
Comment (kyleknap) [inline awscli/customizations/ec2/paginate.py]: Are the ``target_operations`` and ``global_whitelist`` needed? It is really only used for testing and if that is the case why can't you just keep the class variables, instantiate an instance, and then modify the list or dictionary for the test.

Another reason I am asking is that I find the ``target_operations`` interface to be strange. Strange in that it will be a dictionary. I would have more expected that to be a list of operations to inject into and then you would have you class variable ``PER_OPERATION_WHITELIST`` which will add specific parameters to the ``GLOBAL_WHITELIST`` for that operation if it is present in ``PER_OPERATION_WHITELIST``.

File: awscli/customizations/ec2/paginate.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        parsed_args.page_size = self._default_page_size
+
+    def _get_specified_params(self, namespace):
+        attrs = dir(namespace)
+        params = []
+        for attr in attrs:
+            # The default value of generate_cli_skeleton is False where
+            # usually a boolean argument still defaults to None. Therefore
+            # it must be specifically ignored.
+            if attr == 'generate_cli_skeleton':
```
Comment (kyleknap) [inline awscli/customizations/ec2/paginate.py]: Why is this needed? You whitelist it in the ``DEFAULT_GLOBAL_WHITELIST`` right?

File: tests/integration/test_ec2.py [hunk 4]
```diff
@@ -13,23 +13,52 @@
...
+    def setUp(self):
+        self.prefix = 'ec2 describe-instances --region us-west-2'
+
+    def test_describe_instances_with_id(self):
+        command = self.prefix + ' --instance-ids malformed-id'
         result = aws(command)
-        expected_response = ('Request would have succeeded, '
-                             'but DryRun flag is set.')
-        self.assertIn(expected_response, result.stderr)
+        self.assertIn('InvalidInstanceID.Malformed', result.stderr)
```
Comment (kyleknap) [inline tests/integration/test_ec2.py]: Question: If you use MaxResults and provide a malformed instance id, what error gets propogated?  I have the same question for the rest of the test cases. Might be good to add a note or rename the tests as without any background knowledge it is not obvious why we are testing for that specific error, other than for the sake of a quick smoke test of the service.

Especially since for the case, where you have no instances matching the filter, EC2 will actually not complain about MaxResults being present:
```
$ aws ec2 describe-instances --filters Name=instance-id,Values=malformed --max-items 5
{
    "Reservations": []
}

```
That being said I am not sure if there is a great way to test the filter case.

**Comment by kyleknap** [comment]
Approach looks fine.  My main comments were about the interface and questions about the test cases.

File: tests/integration/test_ec2.py (hunk 4 above)
Comment (JordonPhillips) [inline tests/integration/test_ec2.py]: For the id case, the invalid combination error gets thrown. The filters case is harder to be exacting on, but the idea is to make sure it works generally even if the change doesn't impact it.

File: awscli/customizations/ec2/paginate.py (hunk 2 above)
Comment (JordonPhillips) [inline awscli/customizations/ec2/paginate.py]: It seems strange to duplicate that information. I would be worried about future updates missing one location or another. Maybe the naming could be better?

File: awscli/customizations/ec2/paginate.py [hunk 5]
```diff
@@ -14,44 +14,27 @@
...
-            whitelisted for every operation.
-        """
-        self._default_page_size = default_page_size
-        self._target_operations = target_operations
-        if self._target_operations is None:
-            self._target_operations = self.DEFAULT_TARGET_OPERATIONS
-        self._global_whitelist = global_whitelist
-        if self._global_whitelist is None:
-            self._global_whitelist = self.DEFAULT_GLOBAL_WHITELIST
+    DEFAULT_PAGE_SIZE = 1000
```
Comment (kyleknap) [inline awscli/customizations/ec2/paginate.py]: Oh I was fine with the ``page_size`` constructor parameter as it would make sense to individually tune each injector, but I am fine with it being a class variable for now and refactoring it if we need to choose different defaults for different operations

File: tests/integration/test_ec2.py (hunk 4 above)
Comment (kyleknap) [inline tests/integration/test_ec2.py]: I wonder if it is possible if there are any other filter parameter combinations (other than instance-id) that can recreate the regression reliably. Otherwise I am about a zero on adding this test as there is not much value the test adds as even if we reintroduced the regression it would not catch it.

I doubt there will be a different filter parameter combination but it would be good to take a quick look.

File: awscli/customizations/ec2/paginate.py (hunk 5 above)
Comment (JordonPhillips) [inline awscli/customizations/ec2/paginate.py]: It would need refactoring either way since before it was just setting page size. I agree though that there's potential that other operations may need more granularity.

File: awscli/customizations/ec2/paginate.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        'cli_input_json', 'generate_cli_skeleton', 'help', 'max_items',
+        'next_token', 'dry_run', 'starting_token'
+    ]
+
+    DEFAULT_PAGE_SIZE = 1000
+
+    def register(self, event_emitter):
+        """ Register `inject` for each target operation. """
+        event_template = "operation-args-parsed.ec2.%s"
+        for operation in self.TARGET_OPERATIONS.keys():
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: You don't need `.keys()`, you're not mutating the dict.

File: awscli/customizations/ec2/paginate.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    # Parameters which should be whitelisted for every operation.
+    GLOBAL_WHITELIST = [
+        'cli_input_json', 'generate_cli_skeleton', 'help', 'max_items',
+        'next_token', 'dry_run', 'starting_token'
+    ]
+
+    DEFAULT_PAGE_SIZE = 1000
+
+    def register(self, event_emitter):
+        """ Register `inject` for each target operation. """
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: You have leading/trailing spaces on your docstrings which should be removed (pep257)

File: awscli/customizations/ec2/paginate.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            event_emitter.register(event, self.inject)
+
+    def inject(self, event_name, parsed_args, parsed_globals, **kwargs):
+        """ Conditionally inject page_size. """
+        if not parsed_globals.paginate:
+            return
+
+        operation_name = event_name.split('.')[-1]
+
+        whitelisted_params = self.TARGET_OPERATIONS.get(operation_name, None)
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: `,None` is a default value.  I think it's more idiomatic to not provide `None` here (use the default arg only if you have a different value value you want).

File: awscli/customizations/ec2/paginate.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        specified_params = self._get_specified_params(parsed_args)
+
+        for param in specified_params:
+            if param not in whitelisted_params:
+                return
+
+        parsed_args.page_size = self.DEFAULT_PAGE_SIZE
+
+    def _get_specified_params(self, namespace):
+        attrs = dir(namespace)
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: This is not a preferred way to do this.  `vars()` is far more common and used within our code base.  Please use `vars()` here.

File: awscli/customizations/ec2/paginate.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        params = []
+        for attr in attrs:
+            if attr[0] != '_' and getattr(namespace, attr) is not None:
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: This will give you false positives because it's assuming that `None` is the unspecified values.  It doesn't take into account non None default values (i.e booleans that default to False, etc)

File: awscli/handlers.py [hunk 11]
```diff
@@ -142,3 +143,4 @@ def awscli_initialize(event_handlers):
         register_create_keys_from_csr_arguments)
     register_cloudfront(event_handlers)
     register_gamelift_commands(event_handlers)
+    EC2PageSizeInjector().register(event_handlers)
```
Comment (jamesls) [inline awscli/handlers.py]: I'd prefer we stick to the more common pattern of having a single module level function, `register_*`, that handles the registration.  It's by far the common pattern used here, and it will also make it easier to refactor in the future if we start to persue deferred event registration.

File: awscli/customizations/ec2/paginate.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            return
+
+        whitelisted_params += self.GLOBAL_WHITELIST
+        specified_params = self._get_specified_params(parsed_args)
+
+        for param in specified_params:
+            if param not in whitelisted_params:
+                return
+
+        parsed_args.page_size = self.DEFAULT_PAGE_SIZE
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: We should have debug logs that we're messing with the `parsed_args`.  This will make it easier to troubleshoot issues.

File: awscli/customizations/ec2/paginate.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    #    Key:   Operation
+    #    Value: List of parameters to add to whitelist for that operation.
+    TARGET_OPERATIONS = {
+        "describe-volumes": [],
+        "describe-snapshots": ['owner_ids', 'restorable_by_user_ids']
+    }
+
+    # Parameters which should be whitelisted for every operation.
+    GLOBAL_WHITELIST = [
+        'cli_input_json', 'generate_cli_skeleton', 'help', 'max_items',
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: I don't get why some of these are whitelisted.  Why would `help` be a whitelisted parameter?  Same with `--generate-cli-skeleton`.  Why is this relevant for deciding if we inject a page_size param?

**Comment by jamesls** [comment]
I'm not sure that trying to address this at the `parsed_args` layer is going to work.  For example, there's currently a regression here for `--cli-input-json`:

```
$ cat /tmp/foo.json
{
    "SnapshotIds": [
        "snap-05339dbe40d27bfa6"
    ]
}
$ aws ec2 describe-snapshots --cli-input-json file:///tmp/foo.json

An error occurred (InvalidParameterCombination) when calling the DescribeSnapshots operation: The parameter snapshotSet cannot be used with the parameter maxResults
```

It would be ideal if there was some unified way to handle this logic without having to also special case `cli-input-json`.

File: awscli/customizations/ec2/paginate.py (hunk 13 above)
Comment (JordonPhillips) [inline awscli/customizations/ec2/paginate.py]: I added all the generic parameters present in the namespace. I suppose for those two it really doesn't matter since they'll prevent a request anyway.

File: awscli/customizations/ec2/paginate.py (hunk 10 above)
Comment (JordonPhillips) [inline awscli/customizations/ec2/paginate.py]: Modeled params will default to `None` regardless of their type, so we only have to worry about that for custom params that get added to `parsed_args` specifically, not `parsed_globals`. Right now (for these operations) those are `help`, `cli_input_json`, and `generate_cli_skeleton`. Of those, only `generate_cli_skeleton` has a default value. Adding it to the whitelist takes care of that.

Data I have access to might change when I move to a different handler, but at this point I don't have any way to know what the default is.

**Comment by JordonPhillips** [comment]
This is good to look at again

File: tests/unit/customizations/ec2/test_paginate.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        event_name = 'operation-args-parsed.ec2.foo'
+        injector.inject(
+            event_name=event_name,
+            parsed_globals=parsed_globals,
+            call_parameters=call_parameters
+        )
+        page_size = call_parameters.get('PaginationConfig', {}).get('PageSize')
+        self.assertIsNone(page_size)
+
+    def test_global_whitelist(self):
```
Comment (kyleknap) [inline tests/unit/customizations/ec2/test_paginate.py]: How does this test the global whitelist? It looks like it is testing operation level whitelist as this is returning None because ``baz`` is passed as a call_parameter and ``foo`` has an empty whitelist?

**Comment by kyleknap** [comment]
I just had a comment about a test. But otherwise :ship:

File: tests/unit/customizations/ec2/test_paginate.py (hunk 14 above)
Comment (JordonPhillips) [inline tests/unit/customizations/ec2/test_paginate.py]: The only whitelisted parameter is `bar`, any other provided parameters should cause it to not inject page size.

Comment (kyleknap) [inline tests/unit/customizations/ec2/test_paginate.py]: But what specific global whitelist parameter is causing the page size to not being injected? All I see is ``paginate=True``.

Is that what is making it not inject page size? If so  that is really strange when you compare this test and the last because the last test explicitly set paginate to False to get no injection and this one sets paginate to True to get no injection.

Could we just append bar to the current global whitelist and then pass a more explicit parameter in the parsed globals? Furthermore, if page_size is the one causing it to not get injected, it is ambiguous if it is sourced from the global or operation whitelist as your call parameters do not match the operation white list as well.

Comment (JordonPhillips) [inline tests/unit/customizations/ec2/test_paginate.py]: There is no specific parameter that's causing it to not be injected, that's the point of a whitelist: anything not in the list is rejected. `paginate` must be set for anything to be injected regardless of the state of the whitelists.

The global whitelist isn't a whitelist for `parsed_globals` it's a whitelist that is applied globally to all operations. Would it be less confusing to call it universal whitelist?

Comment (kyleknap) [inline tests/unit/customizations/ec2/test_paginate.py]: Oh that makes more sense. Yeah ``GLOBAL_WHITELIST`` definitely through me off because you had global parameters in it before. How do you think ``DEFAULT_PARAMETER_WHITELIST`` sounds? That would be much more clear to me.

File: awscli/customizations/ec2/paginate.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    # Format:
+    #    Key:   Operation
+    #    Value: List of parameters to add to whitelist for that operation.
+    TARGET_OPERATIONS = {
+        "describe-volumes": [],
+        "describe-snapshots": ['OwnerIds', 'RestorableByUserIds']
+    }
+
+    # Parameters which should be whitelisted for every operation.
+    GLOBAL_WHITELIST = ['NextToken', 'DryRun', 'PaginationConfig']
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: Why is `DryRun` in the whitelist?  DryRun is just used for auth checks and doesn't return any data to paginate.  What problem did you run into if you didn't include this in the whitelist?

File: awscli/customizations/ec2/paginate.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        if 'PageSize' in pagination_config:
+            return
+
+        operation_name = event_name.split('.')[-1]
+
+        whitelisted_params = self.TARGET_OPERATIONS.get(operation_name)
+        if whitelisted_params is None:
+            return
+
+        whitelisted_params += self.GLOBAL_WHITELIST
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: Not really a problem because we're relying on the fact that this event is called once, but I find it weird that you're mutating the actual list in self.TARGET_OPERATIONS every time this event handler is called.  This means GLOBAL_WHITELIST gets append to the same list every time this handler is called.

File: awscli/customizations/ec2/paginate.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+        for param in call_parameters:
+            if param not in whitelisted_params:
+                return
+
+        pagination_config['PageSize'] = self.DEFAULT_PAGE_SIZE
+        call_parameters['PaginationConfig'] = pagination_config
+
```
Comment (jamesls) [inline awscli/customizations/ec2/paginate.py]: You can remove this blank line

**Comment by jamesls** [comment]
Had a few questions, but overall approach seems good to me.

File: awscli/customizations/ec2/paginate.py (hunk 16 above)
Comment (JordonPhillips) [inline awscli/customizations/ec2/paginate.py]: Yeah, that should be changed.

File: awscli/customizations/ec2/paginate.py (hunk 15 above)
Comment (JordonPhillips) [inline awscli/customizations/ec2/paginate.py]: There's no problem if it is excluded, but that list is the parameters where there will be no issue putting the paging param in. I'm 0 on keeping it, so I'll remove it if you'd like.

File: tests/integration/test_ec2.py (hunk 4 above)
Comment (JordonPhillips) [inline tests/integration/test_ec2.py]: Those tests are there to verify that the common cases for this operation are covered since we're modifying it. I'll see if I can find a way to reproduce the issue, but either way I think they should be kept.

**Comment by kyleknap** [comment]
Thanks :ship:

**Comment by jamesls** [comment]
:shipit:
